core/invoker/src/main/scala/whisk/core/containerpool/ContainerProxy.scala

```diff
@@ -0,0 +1,388 @@
+            goto(Paused)
+
+        case Event(_: FailureMessage, data: WarmedData) => destroyContainer(data.container)
+        case _ => delay
+    }
+
+    when(Paused, stateTimeout = unusedTimeout) {
+        case Event(job: Run, data: WarmedData) =>
+            implicit val transid = job.msg.transid
+            data.container.resume()
+                .flatMap(_ => run(data.container, job))
+                .map(_ => WarmedData(data.container, job.msg.user.namespace, job.action, Instant.now))
+                .pipeTo(self)
+
+            goto(Running)
+
+        // timeout or removing
+        case Event(StateTimeout | Remove, data: WarmedData) => destroyContainer(data.container)
+    }
+
+    when(Removing) {
+        case Event(job: Run, _) =>
+            // Send the job back to the pool to be rescheduled
+            context.parent ! job
+            stay
+        case Event(ContainerRemoved, _) => stop()
+    }
+
+    // Unstash all messages stashed while in intermediate state
+    onTransition {
+        case _ -> Started => unstashAll()
+        case _ -> Ready   => unstashAll()
+        case _ -> Paused  => unstashAll()
+    }
+
+    initialize()
+
+    /** Delays all incoming messages until unstashAll() is called */
+    def delay = {
+        stash()
+        stay
+    }
+
+    /**
+     * Destroys the container after unpausing it if needed. Can be used
+     * as a state progression as it goes to Removing.
+     *
+     * @param container the container to destroy
+     */
+    def destroyContainer(container: Container) = {
+        context.parent ! ContainerRemoved
+
+        val unpause = stateName match {
+            case Paused => container.resume()(TransactionId.invokerNanny)
+            case _      => Future.successful(())
+        }
+
+        unpause
+            .flatMap(_ => container.destroy()(TransactionId.invokerNanny))
+            .map(_ => ContainerRemoved).pipeTo(self)
+
+        goto(Removing)
+    }
+
+    /**
+     * Runs the job, initialize first if necessary.
+     *
+     * @param container the container to run the job on
+     * @param job the job to run
+     * @return a future completing after logs have been collected and
+     *         added to the WhiskActivation
+     */
+    def run(container: Container, job: Run)(implicit tid: TransactionId): Future[WhiskActivation] = {
+        val actionTimeout = job.action.limits.timeout.duration
+
+        // Only initialize iff we haven't yet warmed the container
+        val initialize = stateData match {
+            case data: WarmedData => Future.successful(Interval.zero)
+            case _                => container.initialize(job.action.containerInitializer, actionTimeout)
+        }
+
+        val activation: Future[WhiskActivation] = initialize.flatMap { initInterval =>
+            val passedParameters = job.msg.content getOrElse JsObject()
+            val boundParameters = job.action.parameters.toJsObject
```

Comment:
I know this is preserving behavior with existing implementation... but this is not necessary because the parameters have already been merged in the controller, so we should be ignoring the action parameters here. Alternatively, the controller should not merge the action parameter (but only merge the package/bindings with the incoming values). I prefer the no remerging here with the invariant that the passed parameters are already fully merged.